### ISSUE TITLE
The issue of illuminating the backside of a mesh after rotating the camera

### ISSUE THREAD

**Issue by JintaoLee-Roger** [vispy/vispy#2525]
Hi, I'm encountering an issue while rendering a surface. After drawing a mesh, I noticed that only the front side is illuminated, and when the camera rotates to the backside, it becomes completely dark. How can I address this problem? (I want to keep the `on_transform_change` function because I have other closed meshes to render as well.)

My code:
```
canvas = scene.SceneCanvas(keys='interactive', bgcolor='w')
view = canvas.central_widget.add_view()
view.camera = scene.TurntableCamera(up='z', fov=60)

# Simple surface plot example
# x, y values are not specified, so assumed to be 0:50
z = np.random.normal(size=(250, 250), scale=200)
z[100, 100] += 50000
z = gaussian_filter(z, (10, 10))
p1 = scene.visuals.SurfacePlot(z=z, color=(0.3, 0.3, 1, 1))
p1.transform = scene.transforms.MatrixTransform()
p1.transform.scale([1/249., 1/249., 1/249.])
p1.transform.translate([-0.5, -0.5, 0])

view.add(p1)

def attach_headlight(mesh, view, canvas):
    light_dir = (0, -1, 0, 0)
    mesh.shading_filter.light_dir = light_dir[:3]
    initial_light_dir = view.camera.transform.imap(light_dir)

    @view.scene.transform.changed.connect
    def on_transform_change(event):
        transform = view.camera.transform
        mesh.shading_filter.light_dir = transform.map(initial_light_dir)[:3]

attach_headlight(p1, view, canvas)
```

front side:

![image](https://github.com/vispy/vispy/assets/38800137/18c82606-e09c-4bfd-bacc-ea7c7f53bee8)

backside:

![image](https://github.com/vispy/vispy/assets/38800137/acba637d-be3f-41cb-ab5a-2186c1702806)

Maybe a function like double-side lighting is my need.

Thanks!

**Comment by djhoese** [comment]
Sorry, I'm a little confused. In your `on_transform_change` function you are explicitly changing the light direction to be on one side of the mesh. If you comment out what you're doing there then it acts as a stationary light that illuminates whatever side is facing the user.

What are you trying to accomplish with the light direction in your transform handler?

Either way, the `mesh.shading_filter` is (if I remember correctly) a convenience for attaching a mesh `ShadingFilter`:

https://vispy.org/api/vispy.visuals.filters.mesh.html#vispy.visuals.filters.mesh.ShadingFilter

I don't think there is a problem (:fearful:) with attaching multiple of these filters if you really wanted double sided lighting with two lights that "rotate" with the object(s)...but I've also never done it before.

**Comment by JintaoLee-Roger** [comment]
1. My needs: I want to draw multiple meshs together. For example, there are two meshs: one is a surface (unclosed mesh), and the other is a tube (closed mesh, approximate).  I hope the side facing the screen is always bright and free from shadows.

2. I set up the `on_transform_change ` for the following reason: as mentioned in this link: (#1479)[https://github.com/vispy/vispy/issues/1479#issue-319621918], I want to illuminate the side of the tube that is facing the screen consistently. To achieve this, I needed the `on_transform_change` function. After adding the `on_transform_change` function, I did indeed achieve this goal, but the surface encountered an issue, it backside is dark.

3. Even without using the `on_transform_change` function, it's still not possible to achieve consistent gloss when rotating the surface, even with well-adjusted lighting direction. For example, in this case, both sides are bright, but the backside appears glossy while the front side doesn't.

4. I cannot add multiple `ShadingFilter` to the same mesh; this seems to be incorrect behavior in VisPy.

**Comment by JintaoLee-Roger** [comment]
Here are some code and images

Without `on_transform_change ` function, both sides of the surface are bright, but the backside appears glossy while the front side doesn't. One side of the tube is bright, but the other side are dark.

```python
import sys
import numpy as np

from vispy import app, scene
from vispy.util.filter import gaussian_filter
from vispy.visuals.filters import ShadingFilter

canvas = scene.SceneCanvas(keys='interactive', bgcolor='w')
view = canvas.central_widget.add_view()
view.camera = scene.TurntableCamera(up='z', fov=60)

z = np.random.normal(size=(250, 250), scale=200)
z[100, 100] += 50000
z = gaussian_filter(z, (10, 10))
p1 = scene.visuals.SurfacePlot(z=z, color=(0.3, 0.3, 1, 1), shading=None)
p1.transform = scene.transforms.MatrixTransform()
p1.transform.scale([1 / 249., 1 / 249., 1 / 249.])
p1.transform.translate([-0.5, -0.5, 0])

x = np.array([0.2] * 10)
y = np.array([0.3] * 10)
zt = np.linspace(-0.5, 0.5, 10)
points = np.c_[x, y, zt]
tube = scene.visuals.Tube(points, 0.1, shading=None)

view.add(p1)
view.add(tube)

filter1 = ShadingFilter(shading='smooth', shininess=200, light_dir=(1, 1, 1))
filter2 = ShadingFilter(shading='smooth', shininess=200, light_dir=(1, 1, 1))
p1.attach(filter1)
tube.attach(filter2)

if __name__ == '__main__':
    canvas.show()
    if sys.flags.interactive == 0:
        app.run()
```

<img width="851" alt="image" src="https://github.com/vispy/vispy/assets/38800137/6972ea5f-f916-4a1f-9ebe-cec7436b1ff2">

**Comment by djhoese** [comment]
I passed `elevation=0` to the turntable camera to kind of "flatten" the x/y/z planes (at least I think) and then set the initial light direction to `(0, 0, -10)` and I think this is "better"? Thoughts?

```python
initial_light_dir = (0, 0, -10)
filter1 = ShadingFilter(shading='smooth', shininess=200, light_dir=initial_light_dir)
filter2 = ShadingFilter(shading='smooth', shininess=200, light_dir=initial_light_dir)
p1.attach(filter1)
tube.attach(filter2)

@view.scene.transform.changed.connect
def on_transform_change(event):
    transform = view.camera.transform
    filter1.light_dir = transform.map(initial_light_dir)[:3]
    filter2.light_dir = transform.map(initial_light_dir)[:3]
```

I'm really not an expert at this stuff so besides playing with a few values I'm not sure I'll be able to help much with the glossiness of the reflection. It looks like you already figured out all of the differences in defaults between Tube and Mesh and have worked around them (ex. setting shading to `None`).

**Comment by djhoese** [comment]
I just released vispy 0.14.0 which includes fixes in #2493 that may affect your results. You might want to try upgrading and see if that makes things look better for you. The conda-forge package should be available in the next hour or so.

**Comment by JintaoLee-Roger** [comment]
Fantastic, thank you for providing the new release; it indeed resolved my problem.
